Add ConnectionStatus component

kinow opened the pull request:
These changes close #93 

Found one way to implement it while working on the websockets issue, so using the momentum to write docs and implement small changes such as this one.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

kinow commented:
Based on the component with the same name, from [Vue UI](https://github.com/vuejs/vue-cli/blob/8fcea225b7b4406ff7c03e34d4e962b45d78ef9f/packages/%40vue/cli-ui/src/components/app/ConnectionStatus.vue).

Instead of running a GraphQL query, we simply check if the WebSocket connection has been established or not. And instead of `div`s with some styles, we use Vuetify's [snackbars](https://vuetifyjs.com/en/components/snackbars).

kinow commented:
Need to write some tests, but here's the component in action.

![2020-01-22 13-26-18 flv](https://user-images.githubusercontent.com/304786/72855291-d5e22380-3d1b-11ea-8b0b-b6e715297b0e.gif)

Note that the UI Server is not automatically started by `jupyterhub`. But once it's up, the Apollo Client should reconnect. Though I noticed that sometimes it appears that the browser would have trouble reconnecting. But nothing wrong in ApolloClient that I could tell. Looks more like it tried so many times that even after focusing on the browser with the mouse, it still ignores.

Opening a new browser normally solves it. If it is confirmed as an issue, this can be tracked separately - still better to have an indicator that the connection is broken I guess?

kinow commented:
Also note that Vue has a lifecycle for pages/components that include things like `beforeDestroy` that are invoked when you refresh a page (never found it in the docs, but quite sure it's invoked).

So if you reload the page, for 1 second you should see the offline notification, then it disappears when page is loaded again.

hjoliver commented:
Nice :+1: 

> Opening a new browser normally solves it. If it is confirmed as an issue, this can be track ed separately - still better to have an indicator that the connection is broken I guess?

IMO, yes.

kinow commented on src/main.js:
```diff
@@ -16,13 +16,23 @@ import router from '@/router'
+let subscriptionClient = null
+if (process.env.NODE_ENV !== 'offline') {
+  subscriptionClient = createSubscriptionClient(graphQLUrls.wsUrl)
+}
```
I tried to avoid having any import from `utils/graphql` here, but unfortunately I found no other way to prevent the ApolloClient of trying to create the WebSocket connection - which is not available in offline mode.

kinow commented on src/utils/graphql.js:
```diff
@@ -4,6 +4,69 @@ import { HttpLink } from 'apollo-link-http'
+export function createSubscriptionClient (wsUrl, options = {}, wsImpl = null) {
+  const opts = Object.assign({
+    reconnect: true,
+    lazy: false
```
We need these two default values. But callers (especially unit tests) are free to change. For unit tests, you want to have `lazy: true`, so that the `SubscriptionClient` doesn't try to create the connection, as node/mocha don't have a WebSocket implementation, resulting in runtime error.

kinow commented:
Ready for review. I've added a changelog entry under the 0.2 release, but that doesn't mean that's the version it will be released.

So no hurry reviewing it. If not 0.2, we can review it for 0.3, and I just amend the commit to move the entry under 0.3, or 0.4, etc :+1: 

To review, take a look at the code & tests. Most of the changes in the `.js` code (not in the `.vue`) are for unit tests. The easiest path, the first which I implemented, was have everything under `createApolloClient()`.

That method was getting bloated, and getting too "smart". It was creating the `ApolloClient`, creating the GraphQL URL's, and now creating a `SubscriptionClient`. The last would result in an error as WebSockets are not available with node/mocha, without a browser or some polyfill.

Mocking ES6 import dependencies is not possible with Sinon. There are other modules like proxyquire that do that, but I thought it was better to change the objects a bit, instead of adding another dependency to the project.

Once running the app, stop the Hub. You should now see a red warning at the top. If you restart the hub, and maybe in another window or even another browser launch the UI server, then that warning should disappear, and both window/browsers should be showing the UI with no issues.

oliver-sanders commented:
Nice.

Works as expected.
